pallets/dkg-metadata/src/lib.rs

```diff
@@ -737,150 +740,172 @@ impl<T: Config> Pallet<T> {
-		const RECENTLY_SENT: &str = "Already submitted a key in this session";
-		let submit_at = submit_at_ref.get::<T::BlockNumber>();
-
-		if let Ok(Some(submit_at)) = submit_at {
-			if block_number < submit_at {
-				frame_support::log::debug!(target: "dkg", "Offchain worker skipping public key submmission");
-				return Ok(())
+		let mut lock = StorageLock::<Time>::new(AGGREGATED_PUBLIC_KEYS_AT_GENESIS);
+		{
+			let _guard = lock.lock();
+			let mut agg_key_ref = StorageValueRef::persistent(AGGREGATED_PUBLIC_KEYS_AT_GENESIS);
+			let mut submit_at_ref = StorageValueRef::persistent(SUBMIT_GENESIS_KEYS_AT);
+			const RECENTLY_SENT: &str = "Already submitted a key in this session";
+			let submit_at = submit_at_ref.get::<T::BlockNumber>();
+
+			if let Ok(Some(submit_at)) = submit_at {
+				if block_number < submit_at {
+					frame_support::log::debug!(target: "dkg", "Offchain worker skipping public key submmission");
+					return Ok(())
+				} else {
+					submit_at_ref.clear();
+				}
 			} else {
-				submit_at_ref.clear();
+				return Err(RECENTLY_SENT)
 			}
-		} else {
-			return Err(RECENTLY_SENT)
-		}
 
-		if !Self::dkg_public_key().1.is_empty() {
-			agg_key_ref.clear();
-			return Ok(())
-		}
+			if !Self::dkg_public_key().1.is_empty() {
+				agg_key_ref.clear();
+				return Ok(())
+			}
 
-		let agg_keys = agg_key_ref.get::<AggregatedPublicKeys>();
+			let agg_keys = agg_key_ref.get::<AggregatedPublicKeys>();
 
-		let signer = Signer::<T, T::OffChainAuthId>::all_accounts();
-		if !signer.can_sign() {
-			return Err(
-				"No local accounts available. Consider adding one via `author_insertKey` RPC.",
-			)
-		}
+			let signer = Signer::<T, T::OffChainAuthId>::all_accounts();
+			if !signer.can_sign() {
+				return Err(
+					"No local accounts available. Consider adding one via `author_insertKey` RPC.",
+				)
+			}
 
-		if let Ok(Some(agg_keys)) = agg_keys {
-			let _res = signer.send_signed_transaction(|_account| Call::submit_public_key {
-				keys_and_signatures: agg_keys.clone(),
-			});
+			if let Ok(Some(agg_keys)) = agg_keys {
+				let _res = signer.send_signed_transaction(|_account| Call::submit_public_key {
+					keys_and_signatures: agg_keys.clone(),
+				});
 
-			agg_key_ref.clear();
-		}
+				agg_key_ref.clear();
+			}
 
-		Ok(())
+			Ok(())
+		}
 	}
 
 	fn submit_next_public_key_onchain(block_number: T::BlockNumber) -> Result<(), &'static str> {
-		let mut agg_key_ref = StorageValueRef::persistent(AGGREGATED_PUBLIC_KEYS);
-		let mut submit_at_ref = StorageValueRef::persistent(SUBMIT_KEYS_AT);
-		const RECENTLY_SENT: &str = "Already submitted a key in this session";
-		let submit_at = submit_at_ref.get::<T::BlockNumber>();
-
-		if let Ok(Some(submit_at)) = submit_at {
-			if block_number < submit_at {
-				frame_support::log::debug!(target: "dkg", "Offchain worker skipping public key submmission");
-				return Ok(())
+		let mut lock = StorageLock::<Time>::new(AGGREGATED_PUBLIC_KEYS);
```

Comment:
Same as above. there should be a dedicated key just for the `submit_next_public_key_onchain` function. for example it should be like the following:
```suggestion
		let mut lock = StorageLock::<Time>::new(SUBMIT_NEXT_PUBLIC_KEY_ON_CHAIN_LOCK);
```